tests/test_notebooks.py

```diff
@@ -0,0 +1,82 @@
+import glob
+import os
+import traceback
+
+import pytest
+from testbook import testbook
+
+import composer
+
+NB_PATH = '../notebooks/up_and_running_with_composer.ipynb'
+nb_root = os.path.join(os.path.dirname(composer.__file__), '..', 'notebooks')
+
+NOTEBOOKS = [
+    os.path.join(nb_root, nb) \
+    for nb in glob.glob(os.path.join(nb_root, '*.ipynb')) \
+]
+
+
+def seek(tb, tag):
+    """ generator that seeks next cell with the metadata tag,
+    and returns the cell index.
+    """
+    for idx, cell in enumerate(tb.cells):
+        metadata = cell['metadata']
+        if 'tags' in metadata and tag in metadata['tags']:
+            yield idx
+
+
+@pytest.fixture(params=NOTEBOOKS)
+def tb(request):
+    notebook_path = request.param
+    with testbook(notebook_path) as tb:
+        yield tb
+
+
+def _assert_tags_correct(tb):
+    """Checks that the trainer_fit tag exists
+       whenever trainer.fit is called.
+    """
+    for cell in tb.cells:
+        if cell['cell_type'] == 'code' and 'trainer.fit' in cell['source']:
+            metadata = cell['metadata']
+            if 'tags' not in metadata or 'trainer_fit' not in metadata['tags']:
+                raise ValueError("In cell that starts with '{}', trainer.fit was found "
+                                 "but the trainer_fit tag was not in metadata.".format(cell['source'][:40]))
+
+
+@pytest.mark.timeout(120)
+@pytest.mark.notebooks
+def test_notebook(tb):
+    _assert_tags_correct(tb)
+
+    start = 0
+    for stop in seek(tb, tag="trainer_fit"):
+        # run until right before the next trainer.fit call
+        tb.execute_cell(range(start, stop))
+
+        try:
+            # patch the trainer to only train for 5 batches
+            tb.inject("""
+                from composer.core import Time
+                trainer.state.max_duration = Time.from_timestring('1ep')
+                trainer.state.train_subset_num_batches = 5
```

Comment:
Will we want to monkeypatch the dataloader to load synthetic data? Perhaps in each notebook we can have a cell that defines `synthetic = False` by default, but it has code-paths for `synthetic = False` and `synthetic = True`? I could also see a user wanting to run on synthetic data if they don't want to download imagenet for example.

Reply:
Another option could be manually override the `dataset` argument in the test case before the dataloader is constructed, if we don't want to expose the synthetic data option to the user.

Reply:
I also think we may need to override the dataloader `num_workers` to be `0` and `pin_memory` to be false. I remember running into issues when configuring jenkins where the test pod would appear to randomly crash, but setting the dataloader num workers to 0 fixed it.

Reply:
We could mock the dataloader, but since some notebooks may want to plot the output data, or do other types of analysis, this is a slippery slope of features and complexity that I do not think we need just yet. Overriding the dataset would need knowledge of each individual notebooks.

After this investigation, my conclusion that we will have to live with daily integration tests of the notebooks and potentially flaky notebooks on `dev` branch. Public release notebooks will always work, as they are tied to a specific tag of `composer` and tested during release.